src/lib.rs

```diff
@@ -91,15 +91,46 @@ pub use crate::prepare::LevelRun;
 pub use crate::char_data::{bidi_class, HardcodedBidiData};
 
 use alloc::borrow::Cow;
+use alloc::boxed::Box;
 use alloc::string::String;
 use alloc::vec::Vec;
 use core::cmp;
 use core::iter::repeat;
 use core::ops::Range;
+use core::str::CharIndices;
 
 use crate::format_chars as chars;
 use crate::BidiClass::*;
 
+/// Trait that abstracts over a text source for use by the bidi algorithms.
+/// We implement this for str (UTF-8) and for [u16] (UTF-16, native-endian).
+pub trait TextSource {
+    /// Return the length of the text in code units.
+    fn len(&self) -> usize;
+
+    /// Get the character at a given code unit index, along with its length in code units.
+    /// Returns None if index is out of range, or points inside a multi-code-unit character.
+    /// Returns REPLACEMENT_CHARACTER for any unpaired surrogates in UTF-16.
+    fn char_at(&self, index: usize) -> Option<(char, usize)>;
+
+    /// An iterator over the text returning Unicode characters,
+    /// REPLACEMENT_CHAR for invalid code units.
+    fn chars(&self) -> Box<dyn DoubleEndedIterator<Item = char> + '_>;
```

Comment:
issue: please no allocation or dynamic dispatch for this kind of thing


Instead, please add associated types for iterators. Eventually we can use type-alias-impl-trait.

Also I don't want to commit to this API just yet. I suggest is making this a sealed trait with `doc(hidden)` methods, and documenting it as unstable for implementation or direct use.